kernel/init.c

```diff
@@ -280,36 +316,67 @@ static void z_sys_init_run_level(enum init_level level)
 		const struct device *dev = entry->dev;
 
 		if (dev != NULL) {
-			int rc = 0;
-
-			if (entry->init_fn.dev != NULL) {
-				rc = entry->init_fn.dev(dev);
-				/* Mark device initialized. If initialization
-				 * failed, record the error condition.
-				 */
-				if (rc != 0) {
-					if (rc < 0) {
-						rc = -rc;
-					}
-					if (rc > UINT8_MAX) {
-						rc = UINT8_MAX;
-					}
-					dev->state->init_res = rc;
-				}
-			}
+#ifdef CONFIG_DEVICE_DEFER_INIT
+			if (device_before_init_registered_cb != NULL) {
```

Comment:
How about using an iterable section of callbacks instead of a single one? The advantage is that the callbacks are registered at build time (no need to have SYS_INIT to register it) and you can have more than one of them. To implement that you'd have to define a struct with the callback as a member, then you can declare it in the linker with `ITERABLE_SECTION_ROM`, in the code with `STRUCT_SECTION_ITERABLE` and then iterate through them here with `STRUCT_SECTION_FOREACH`. I think it'd be fair to iterate through all callbacks for all devices, I'd expect the callback implementation to start with something like `if (dev != <statically defined device>) {return DEVICE_INIT;}` so there should not be much overhead.